pulsar-client/src/main/java/org/apache/pulsar/client/impl/TransactionMetaStoreHandler.java

```diff
@@ -335,41 +402,77 @@ void handleEndTxnResponse(CommandEndTxnResponse response) {
             }
             op.callback.complete(null);
         } else {
-            LOG.error("Got end txn response for request {} error {}", response.getRequestId(), response.getError());
             handleTransactionFailOp(response.getError(), response.getMessage(), op);
+            if (response.getError() == ServerError.TransactionCoordinatorNotFound) {
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug("Get a response for request {} error TransactionCoordinatorNotFound and try it again",
+                            response.getRequestId());
+                }
+                timer.newTimeout(timeout -> {
+                    long requestId = client.newRequestId();
+                    TxnID txnID = op.txnID;
+                    TxnAction action = op.action;
+                    BaseCommand cmd = Commands.newEndTxn(requestId, txnID.getLeastSigBits(),
```

Comment:
should cmd should retain, otherwise it will be release in `cnx().ctx().writeAndFlush(op.cmd, cnx().ctx().voidPromise());`